run.py

```diff
@@ -197,13 +201,160 @@ def get_files(args, dirname, min_size=1, min_date=''):
+                    os.rename(input_fname, input_fname.replace(".gz", ".fq.gz"))
+
+                # Add .fq extensions to input files
+                inputs = [i.replace(".gz", ".fq.gz") for i in inputs]
+
+                # Get subset of inputs
+                subsets, total_reads, subset_reads  = first_subset_fastq(inputs, subset_size)
+                subset_reads_dict[inputs[0]] = subset_reads
+                total_reads_dict[inputs[0]] = total_reads
+                
+                # Compute average read lengths. For paired-end reads, average length is
+                # computed only from pair1 reads.
+                print("Calculating average read length...")
+                def calculate_average_read_length(file_path):
+                    total_len = 0
+                    total_reads = 0
+                    with open(file_path, 'rt') as inf:
+                        for (title, sequence, quality) in FastqGeneralIterator(inf):
+                            total_len += len(sequence)
+                            total_reads += 1
+                    return round(total_len / total_reads)
+
+                avg_length = calculate_average_read_length(subsets[0])
+                print("Done. Average read length is ", avg_length)
+
+                ribodetector_cmd = [
+                    "ribodetector_cpu",
+                    "--ensure", "rrna",
+                    "--threads", "28"
+                    ]
+                ribodetector_cmd.extend(["--len", str(avg_length)])
+
+                ribodetector_cmd.append("--input")
+                ribodetector_cmd.extend(subsets)
+
+                # RiboDetector outputs fastq files containing non-rRNA sequences
+                # https://github.com/hzi-bifo/RiboDetector
+                ribodetector_cmd.append("--output")
+                ribodetector_cmd.extend(tmp_fq_outputs)
+
+                subprocess.check_call(ribodetector_cmd)
+
+                # Count number of rRNA reads in subset
+                non_rrna_count = sum(1 for _ in FastqGeneralIterator(open(tmp_fq_outputs[0], 'rt')))
+                rrna_reads_dict[inputs[0]] = subset_reads - non_rrna_count
+
+        
+        # Calculate the total number of reads across all inputs in sample
+        total_reads_all = sum(total_reads_dict.values())
+
+        # Calculate the weighted fraction for each input
+        weighted_fractions = []
+        for input_filename in total_reads_dict:
+            fraction_rrna_in_subset = rrna_reads_dict[input_filename] / subset_reads_dict[input_filename]
+            weighted_fraction = fraction_rrna_in_subset * total_reads_dict[input_filename]
+            weighted_fractions.append(weighted_fraction)
+
+        # Calculate the weighted average fraction of rRNA reads across all inputs in sample
+        total_weighted_fraction = sum(weighted_fractions)
+        weighted_rrna_fraction = total_weighted_fraction / total_reads_all
```

Comment:
You can use https://numpy.org/doc/stable/reference/generated/numpy.average.html to compute a weighted average

Reply:
Updated